### PR TITLE
docs: workspace filtering

### PR DESCRIPTION
**Pull request by zkochan**
Unresolved questions:

- [ ] should we support the selection of only directory dependencies/dependents?

### CODE REVIEW

File: workspace/filtering.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+pnpm recursive install -- foo bar @pnpm/config
+```
+
+### Selecting a package with dependencies
+
+To select a package and its dependencies, prefix the package name with two dots: `..<pkg name>`.
+For instance, the next command will run installation in all dependencies of `foo` and in `foo`:
+
+```
+pnpm recursive install -- ..foo
```
Comment (zkochan) [inline workspace/filtering.md]: @vjpr thinks it should be the other way around. This is dependents and `foo..` is dependencies. Who else thinks this should be vice versa?

**Comment by vjpr** [comment]
`..bar` should mean include all dependents
`bar..` should mean include all dependencies

**Why?** 

We most commonly think of dependency chains like:

`foo` depends on `bar` which depends on `baz`

Yarn represents dependency chains like this in its logging: `foo > bar > baz`

If we are `bar`, then `..bar` = `foo` and `bar..` = `baz`. Assuming we think the ellipsis means that we want to include more in the direction of the ellipsis, like its meaning something like "etc."

---

Also, a reason not to use the syntax (or to support flags too) is that in zsh `—scope={foo,bar}` expands to `—scope=foo —scope=bar`.

So something like `pnpm multi test --scope={$(find-packages-matching-some-criteria)}` is much harder (more inconvenient) to do.

**Comment by zkochan** [comment]
@pgonzal, @DanielRuf, @huochunpeng, @andreineculau maybe you'll have some suggestions for this specs

**Comment by DanielRuf** [comment]
I'll answert tomorrow. =)

**Comment by DanielRuf** [comment]
I'll answer tomorrow. =)

**Comment by zkochan** [comment]
I think I agree that `foo..` should be dependencies. I'll update the specs.

Also, just noting, `foo..` means all dependencies and subdependencies. So not only the direct dependencies.

**Comment by vjpr** [comment]
> Also, just noting, foo.. means all dependencies and subdependencies. So not only the direct dependencies.

Is it worth having a special syntax for direct deps?

Also what about `foo>...`. Is similar to how we talk display deps `bar > foo > baz`.

**Comment by zkochan** [comment]
I am not sure there is a usecase for direct deps. But if we want to leave the possibility to support it, we can make dependencies three dots `foo...`. Then in future we may make two dots for direct dependencies.

```
foo.. - direct
foo... - all
```

to me the `foo>..,` looks as confusing as `foo...` but the 2nd makes less noise

**Comment by vjpr** [comment]
> to me the foo>.., looks as confusing as foo... but the 2nd makes less noise

Agree.

---

How will filtering by globbing work? Will the same syntax be used?

Or separate `ignore` switch: `pnpm multi link --ignore=package-group/foos/**`

**Comment by zkochan** [comment]
I guess it would work but I am not 100% sure it would still be needed, that is why I am not including it in the specs for now. There is the possibility to select by path and by names. If some grouping is needed, folder structure can be used

**Comment by DanielRuf** [comment]
> foo.. - direct
> foo... - all

I also agree, this is the better approach regarding the dependencies.

**Comment by DanielRuf** [comment]
So far looks good and should be very useful to restrict updates (other package managers update too much).

**Comment by zkochan** [comment]
What exactly do you mean by "restrict updates"? Updates to these specs?

>>   foo.. - direct
>>    foo... - all
>
> I also agree, this is the better approach regarding the dependencies.

You believe this would be a useful feature? I personally don't know if I would ever need to select only direct deps. I will update the spec though, to use the 3 dots. We'll see if users will ask for this feature or not

**Comment by DanielRuf** [comment]
> What exactly do you mean by "restrict updates"? Updates to these specs?
> 
> foo.. - direct
> foo... - all
> 
> I also agree, this is the better approach regarding the dependencies.

No, if we have the two and three dot solution to tell pnpm which dependencies (direct or all) should be installed / updated and so on. A bit more granular and exact then current solutions.

**Comment by zkochan** [comment]
So you think it would be useful to be able to run `pnpm recursive update -- foo..`?

In a monorepo like this:

```
packages
|- foo
|- bar
|- qar
```

with this relations `foo > bar > qar`

`pnpm recursive update -- foo..` will mean basically

```
pnpm update --prefix packages/foo & pnpm update --prefix packages/bar
```

but it it won't affect how deep the update will go inside foo/bar. That is controlled by the `--depth` option

**Comment by DanielRuf** [comment]
Exactly. But besides this, the suggested RFC for workspace filtering by path and package name is definitely a good idea and seems helpful.

**Comment by zkochan** [comment]
It has been on review for a week, so I am merging the specs.

**Pull request merged**